📚 DOCS: Add v2.0.0b1 changelog (#5426)
aiidateam/aiida-core: CHANGELOG.md

  - Note, to make full use of object compression, one should periodically run `verdi storage maintain`.
  - See the [repository design section](docs/source/internals/storage/repository.rst) for details.

- Command-line interaction with a profile's storage has been moved from `verdi database` to `verdi storage`.

- The AiiDA archive format has been redesigned as the `sqlite_zip` storage backend.
  - See the [`sqlite_zip` storage format](docs/source/internals/storage/sqlite_zip.rst) for details.
  - The new format allows for streaming of data during exports and imports, significantly reducing both the time and memory utilisation of these actions.
  - The archive can now be loaded directly as a (read-only) profile, without the need to import it first, see [this Jupyter Notebook tutorial](docs/source/howto/archive_profile.md).

On first using `aiida-core` v2.0, your AiiDA configuration will be automatically migrated to the new version (this can be reverted by `verdi config downgrade`).

> To update existing profiles and archives to the new storage formats, simply use `verdi storage migrate`, or `verdi archive migrate`.

Reviewer: ```suggestion
To update existing profiles and archives to the new storage formats, simply use `verdi storage migrate` and `verdi archive migrate`, respectively.
```

:::{important}
The migration of large storage repositories is a potentially time-consuming process.
It may take several hours to complete, depending on the size of the repository.

> It is also advisable to make a full manual backup of any AiiDA setup with important data: see [the installation management section](docs/source/howto/installation.rst) for more information.

Reviewer: Why not refer directly to the backup section?

Author: So the idea here, is that the Markdown links will work, when you are viewing the "raw" changelog file.

i.e. if you go to https://github.com/chrisjsewell/aiida_core/blob/v2-changelog/CHANGELOG.md

<img width="1034" alt="image" src="https://user-images.githubusercontent.com/2997570/158086011-2d6e3423-1fba-46e2-a7a1-2330355cb5bc.png">

You can click on the link, and it will actually take you to that documentation source page.

That would not work, if you simply use e.g. `` {ref}`the installation management section <how-to:installation:backup>` ``

In MyST you can actually link to heading anchors: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors,
BUT that only works if the "target" page is also Markdown,
e.g. `[the installation management section](docs/source/howto/installation.md#backing-up-your-installation)`

- `Node.open`: only mode `r` and `rb` are supported, [use `put_object_from_` methods instead](https://github.com/aiidateam/aiida-core/issues/4721#issuecomment-920100415)
- `Node.get_object_content`: only mode `r` and `rb` are supported
- `Node.put_object_from_tree`: argument `contents_only` was removed
- `Node.put_object_from_tree`: argument `force` was removed
- `Node.put_object_from_file`: argument `force` was removed
- `Node.put_object_from_filelike`: argument `force` was removed
- `Node.delete_object`: argument `force` was removed

Added:

- `Node.walk`
- `Node.copy_tree`
- `Node.is_valid_cache` setter
- `Node.objects.iter_repo_keys`

Additionally, `Node.open` should always be used as a context manager, for example:

```python
with node.open('filename.txt') as handle:
    content = handle.read()
```

:::

#### QueryBuilder

When using the {class}`~aiida.orm.QueryBuilder` to query the database, the following changes have been made:

- The `Computer`'s `name` field is now replaced with `label` (as previously deprecated in v1.6)
- The `QueryBuilder.queryhelp` attribute is deprecated, for the `as_dict` (and `from_dict`) methods

> - The `QueryBuilder.first` method now allows the `flat` argument, which will return a single item, instead of a list of one item

Reviewer: ```suggestion
- The `QueryBuilder.first` method now allows the `flat` argument, which will return a single item, instead of a list of one item, if only a single projection is defined.
```

For example:

```python
from aiida.orm import QueryBuilder, Computer
query = QueryBuilder().append(Computer, filters={'label': 'localhost'}, project=['label']).as_dict()
QueryBuilder.from_dict(query).first(flat=True)  # -> 'localhost'
```

For further information, see [How to find and query for data](docs/source/howto/query.rst).

#### Dict usage

> The {class}`~aiida.orm.Dict` class has been updated to support more native `dict` behaviour:

Reviewer: Unless you have added this elsewhere, would be good to add that the explicit `dict` keyword in the constructor is no longer required.

Author: added below

> - Equality (`==`/`!=`) comparisons now compare the dictionaries, rather than the UUIDs

Author: ```suggestion
- Initialisation can now use `Dict({'a': 1})`, instead of `Dict(dict={'a': 1})`. This is also the case for `List([1, 2])`.
- Equality (`==`/`!=`) comparisons now compare the dictionaries, rather than the UUIDs
```

: `verdi storage maintain` has also been added, to allow for maintenance of the storage, for example, to optimise the storage size.

`verdi archive version` and `verdi archive info` (replace `verdi archive inspect`)
: This change synchronises the commands with the new `verdi storage version` and `verdi storage info` commands.

`verdi group move-nodes`
: This command moves nodes from a source group to a target group (removing them from one and adding them to the other).

`verdi code setup`
: There is a small change to the order of prompts, in interactive mode.
: The uniqueness of labels is now validated, for both remote and local codes.

`verdi code test`
: Run tests for a given code to check whether it is usable, including whether remote executable files are available.

See [AiiDA Command Line](docs/source/reference/command_line.rst) for more information.

### Development improvements

The build tool for `aiida-core` has been changed from `setuptools` to [`flit`](https://github.com/pypa/flit).
This allows for the project metadata to be fully specified in the `pyproject.toml` file, using the [PEP 621](https://www.python.org/dev/peps/pep-0621) format.

> Note, development installs of `aiida-core` now require `pip>=21`.

Reviewer: ```suggestion
Note, editable installs (using the `-e` flag for `pip install`) of `aiida-core` now require `pip>=21`.
```
